runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -30,44 +30,198 @@
 #include "chpltypes.h"
 #include "error.h"
 
-void chpl_mem_layerInit(void) {
-  void* start;
-  size_t size;
 
-  chpl_comm_desired_shared_heap(&start, &size);
+//TODO move the chunk replacement code into a separate file
 
+// TODO put these into a struct?
+void*  heap_base = NULL;
+size_t heap_size = 0;
+size_t cur_heap_offset = 0;
+
+// *** Chunk hook replacements *** //
+// See http://www.canonware.com/download/jemalloc/jemalloc-latest/doc/jemalloc.html#arena.i.chunk_hooks
+
+// Our chunk replacement hook for allocations. Grab memory out of the shared
+// heap and give it to jemalloc.
+static void* chunk_alloc(void *chunk, size_t size, size_t alignment, bool *zero, bool *commit, unsigned arena_ind) {
+
+  void* p = NULL;
+  size_t p_offset;
+
+  // TODO cleanup some of the pointer arithmetic. The current code was
+  // originally copied out of the tcmalloc shim, and could use a little cleanup
+
+  // compute our current pointer into the shared heap based on the current
+  // offset aligned to "alignment"
+  p = (void*)((((uintptr_t) heap_base + cur_heap_offset + alignment - 1) / alignment) * alignment);
+
+  // """
+  // If chunk is not NULL, the returned pointer must be chunk on success or
+  // NULL on error
+  // """
+  if (chunk && chunk != p) {
+    return NULL;
+  }
+
+  p_offset = (uintptr_t)p - (uintptr_t)(heap_base);
+
+  size = ((size + alignment - 1) / alignment) * alignment;
+
+  // If we're out of space on the heap, return NULL
+  if (size > heap_size - p_offset) {
+    return NULL;
+  }
+
+  // Update the current pointer, now that we've past any early returns.
+  cur_heap_offset = p_offset + size;
+
+  // Zero if we were told to
+  if (*zero) {
```

Comment:
I'm not sure we need to do this.  Does jemalloc require zeroed memory for its own purposes? We don't -- we don't make any guarantees about the contents of malloc'ed space, and of course we zero calloc'ed space ourselves.
Separately, I think this would be where we'd set NUMA locality (because we'll likely set locality just by touching new heap pages), so it might change in the future, but for now if we need to keep it here, it's fine. (If this zeroing stays here, you might add a note that development of NUMA locality support should start by looking here.)